src/rpcmasternode.cpp

```diff
@@ -601,5 +608,222 @@ UniValue masternodelist(const UniValue& params, bool fHelp)
+                ssMnb << mnb;
+                statusObj.push_back(Pair("hex", HexStr(ssMnb.begin(), ssMnb.end())));
+            } else {
+                failed++;
+                statusObj.push_back(Pair("errorMessage", errorMessage));
+            }
+
+            resultsObj.push_back(Pair("status", statusObj));
+        }
+        pwalletMain->Lock();
+
+        UniValue returnObj(UniValue::VOBJ);
+        returnObj.push_back(Pair("overall", strprintf("Successfully created broadcast messages for %d masternodes, failed to create %d, total %d", successful, failed, successful + failed)));
+        returnObj.push_back(Pair("detail", resultsObj));
+
+        return returnObj;
+    }
+
+    if (strCommand == "decode")
+    {
+        if (params.size() != 2)
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Correct usage is 'masternodebroadcast decode \"hexstring\"'");
+
+        CMasternodeBroadcast mnb;
+
+        if (!DecodeHexMnb(mnb, params[1].get_str()))
+            throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Masternode broadcast message decode failed");
+
+        if(!mnb.VerifySignature())
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Masternode broadcast signature verification failed");
+
+        UniValue resultObj(UniValue::VOBJ);
+
+        resultObj.push_back(Pair("vin", mnb.vin.ToString()));
+        resultObj.push_back(Pair("addr", mnb.addr.ToString()));
+        resultObj.push_back(Pair("pubkey", CBitcoinAddress(mnb.pubkey.GetID()).ToString()));
+        resultObj.push_back(Pair("pubkey2", CBitcoinAddress(mnb.pubkey2.GetID()).ToString()));
+        resultObj.push_back(Pair("vchSig", EncodeBase64(&mnb.vchSig[0], mnb.vchSig.size())));
+        resultObj.push_back(Pair("sigTime", mnb.sigTime));
+        resultObj.push_back(Pair("protocolVersion", mnb.protocolVersion));
+        resultObj.push_back(Pair("nLastDsq", mnb.nLastDsq));
+
+        UniValue lastPingObj(UniValue::VOBJ);
+        lastPingObj.push_back(Pair("vin", mnb.lastPing.vin.ToString()));
+        lastPingObj.push_back(Pair("blockHash", mnb.lastPing.blockHash.ToString()));
+        lastPingObj.push_back(Pair("sigTime", mnb.lastPing.sigTime));
+        lastPingObj.push_back(Pair("vchSig", EncodeBase64(&mnb.lastPing.vchSig[0], mnb.lastPing.vchSig.size())));
+
+        resultObj.push_back(Pair("lastPing", lastPingObj));
+
+        return resultObj;
+    }
+
+    if (strCommand == "relay")
+    {
+        if (params.size() != 2)
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Correct usage is 'masternodebroadcast relay \"hexstring\"'");
+
+        CMasternodeBroadcast mnb;
+
+        if (!DecodeHexMnb(mnb, params[1].get_str()))
+            throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Masternode broadcast message decode failed");
+
+        if(!mnb.VerifySignature())
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Masternode broadcast signature verification failed");
+
+        mnodeman.UpdateMasternodeList(mnb);
+        mnb.Relay();
```

Comment:
I think we should use something like mnb.CheckInputsAndAdd(nDos) instead of UpdateMasternodeList + Relay because we should really check the mnb before relay to network. :sunglasses: